R/PipeOpEncodePL.R

```diff
@@ -0,0 +1,409 @@
+#'
+#' @section State:
+#' The `$state` is a named `list` with the `$state` elements inherited from [`PipeOpEncodePL`]/[`PipeOpTaskPreproc`].
+#'
+#' @section Parameters:
+#' The parameters are the parameters inherited from [`PipeOpTaskPreproc`], as well as:
+#' * `numsplits` :: `integer(1)` \cr
+#'   Number of bins to create. Default is `2`.
+#' * `type` :: `integer(1)`\cr
+#'   Method used to calculate sample quantiles. See help of [`stats::quantile`]. Default is `7`.
+#'
+#' @section Internals:
+#' This overloads the `private$.get_bins()` method of [`PipeOpEncodePL`] and uses the [`stats::quantile`] function
+#' to derive the bins used for piecewise linear encoding.
+#'
+#' @section Fields:
+#' Only fields inherited from [`PipeOp`].
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpEncodePL`]/[`PipeOpTaskPreproc`]/[`PipeOp`].
+#'
+#' @references
+#' `r format_bib("gorishniy_2022")`
+#'
+#' @family PipeOps
+#' @family Piecewise Linear Encoding PipeOps
+#' @template seealso_pipeopslist
+#' @include PipeOpTaskPreproc.R
+#' @export
+#' @examples
+#' library(mlr3)
+#'
+#' task = tsk("iris")$select(c("Petal.Width", "Petal.Length"))
+#' pop = po("encodeplquantiles")
+#'
+#' train_out = pop$train(list(task))[[1L]]
+#' # Calculated bin boundaries per feature
+#' pop$state$bins
+#' # Each feature was split into two encoded features using piecewise linear encoding
+#' train_out$head()
+#'
+#' # Prediction works the same as training, using the bins learned during training
+#' predict_out = pop$predict(list(task))[[1L]]
+#' predict_out$head()
+#'
+#' # Binning into three bins per feature
+#' # Using the nearest even order statistic for caluclating quantiles
+#' pop$param_set$set_values(numsplits = 4, type = 3)
+#'
+#' train_out = pop$train(list(task))[[1L]]
+#' # Calculated bin boundaries per feature
+#' pop$state$bins
+#' # Each feature was split into three encoded features using piecewise linear encoding
+#' train_out$head()
+PipeOpEncodePLQuantiles = R6Class("PipeOpEncodePLQuantiles",
+  inherit = PipeOpEncodePL,
+  public = list(
+    initialize = function(id = "encodeplquantiles", param_vals = list()) {
+      ps = ps(
+        numsplits = p_int(lower = 2, default = 2, tags = c("train", "predict")),
+        type = p_int(lower = 1, upper = 9, default = 7, tags = c("train", "predict"))
+      )
+      super$initialize(id, param_set = ps, param_vals = param_vals, packages = "stats")
+    }
+  ),
+  private = list(
+
+    .get_bins = function(task, cols) {
+      numsplits = self$param_set$values$numsplits %??% 2
+      # Defaulting to default value in stats::quantile, i.e. method 7
+      type = self$param_set$values$type %??% 7
+
+      lapply(task$data(cols = cols), function(d) {
+        unique(c(
+          min(d, na.rm = TRUE),
+          stats::quantile(d, seq(1, numsplits - 1) / numsplits, na.rm = TRUE, names = FALSE, type = type),
+          max(d, na.rm = TRUE)
```

Comment:
```suggestion
        stats::quantile(d, seq(0, 1, length.out = numsplits + 1), na.rm = TRUE, names = FALSE, type = type),
```

please make tests that this works for all the different 'type's

Reply:
in particular that it includes mins and max